### PR TITLE
Zeitwerk

### PR DESCRIPTION
**Pull request by ofedoren**
Based on https://github.com/theforeman/foreman/pull/9148.

### CODE REVIEW

**Comment by ofedoren** [comment]
I guess it's open for discussions/suggestions :) I've checked that the applications starts with this PR (couldn't test all the functionality since my setup contains a lot of plugins, which are not Zeitwerk compatible and due to DB changes some pages aren't available), so I'd say overall we're fine after those changes. The main issue is that it's a really breaking change and plugins must be fixed in time.


I'll ping here all the people who touched this area just in case: @ezr-ondrej, @stejskalleos, @adamruzicka. 

UPD: I hope someone will give me a hint how we (or me) could fix the tests since there is not much information from Jenkins and locally almost all the tests are broken, but in a different way... :/

File: app/services/setting_registry.rb [hunk 1]
```diff
@@ -132,18 +132,9 @@ def load_definitions
...
-      @known_categories = @known_descendants.map(&:name) << 'Setting'
-      @values_loaded_at = nil # force all values to be reloaded
-    end
-    @known_categories
-  end
-
   def load_values(ignore_cache: false)
     # we are loading only known STIs as we load settings fairly early the first time and plugin classes might not be loaded yet.
-    settings = Setting.unscoped.where(category: known_categories).where.not(value: nil)
+    settings = Setting.unscoped.where(category: 'Setting').where.not(value: nil)
```
Comment (ofedoren) [inline app/services/setting_registry.rb:137]: My main concern is here. Not sure if we can do this right now, any ideas @ezr-ondrej? :)

File: config/application.rb [hunk 2]
```diff
@@ -115,7 +115,7 @@ class Application < Rails::Application
     config.autoload_paths += %W(#{config.root}/app/models/compute_resources)
     config.autoload_paths += %W(#{config.root}/app/models/fact_names)
     config.autoload_paths += %W(#{config.root}/app/models/lookup_keys)
-    config.autoload_paths += %W(#{config.root}/app/models/host_status)
+    # config.autoload_paths += %W(#{config.root}/app/models/host_status)
```
Comment (ofedoren) [inline config/application.rb:118]: This is from previous commits. My guess is we can remove this line completely.

File: config/initializers/zeitwerk.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  'aix' => 'AIX',
+  'nxos' => 'NXOS',
+  'vrp' => 'VRP',
+  'sso' => 'SSO',
+  'puppet_ca_certificate' => 'PuppetCACertificate',
+  'url_resolver' => 'URLResolver',
+  'ztp_record' => 'ZTPRecord',
+  'aaaa_record' => 'AAAARecord',
+  'ptr4_record' => 'PTR4Record',
+  'ptr6_record' => 'PTR6Record'
```
Comment (ofedoren) [inline config/initializers/zeitwerk.rb:28]: This is basically workaround for `/lib` directory. Also, could please someone tell me why do we even have `/lib` in `eager_load_paths`? And also we have there code which is tightly bound to the whole application and per my thinking and Rails suggestions we should move it to `app/lib` directory or somehow refactor.

File: app/services/setting_registry.rb (hunk 1 above)
Comment (ezr-ondrej) [reply app/services/setting_registry.rb:137]: This basically means we will no longer support any other setting then DSL based, which is IMHO time to do.
We did deprecated it https://github.com/theforeman/foreman/blob/a439e1e92e9f98f3dbc8ed2f204bd53cc861565e/app/models/setting.rb#L204

I'd love a separate PR for that, ideally dropping the category column from setting alltogether, for us to be explicit about the fact. I'll be back online on 11th and I'll be happy to take a look into it that week.

File: config/application.rb (hunk 2 above)
Comment (ezr-ondrej) [reply config/application.rb:118]: and all the above IMHO. we should not add subfolders of `app/*` as all the folders are roots for autoloading and if the files follow the naming patterns expected by autoloaders, it should work without this.

File: config/initializers/zeitwerk.rb (hunk 3 above)
Comment (ezr-ondrej) [reply config/initializers/zeitwerk.rb:28]: We (mainly @stejskalleos) has started to work on exactly this and I believe `/lib` is not autoloaded anymore, we were not sure where everywhere we would need to add require statements, but ideal state is `/lib` content being explicitly required where needed (or from `application.rb`) ... and indeed if there are some strictly application related files that belongs to lib, those should be in `app/lib`

File: app/models/compute_resources/foreman/model/vmware.rb [hunk 4]
```diff
@@ -818,3 +817,5 @@ def cachekey_with_cluster(key, cluster_id = nil)
     end
   end
 end
+
+require 'fog_extensions/vsphere/mini_servers'
```
Comment (stejskalleos) [inline app/models/compute_resources/foreman/model/vmware.rb]: Question - why this change? Is there a difference between `require` at the beginning of file and at the end of the file?

File: app/services/setting_registry.rb (hunk 1 above)
Comment (stejskalleos) [reply app/services/setting_registry.rb:137]: @ezr-ondrej I don't mind take care about deprecating the Settings, but I don't have much idea where to start ... If you point me to what needs to be done I'll push the PR

**Comment by stejskalleos** [comment]
[test foreman]

File: app/models/compute_resources/foreman/model/vmware.rb (hunk 4 above)
Comment (ofedoren) [reply app/models/compute_resources/foreman/model/vmware.rb]: This is weird... At the time I've been writing this, I've faced circular dependency in loading `Vmware` constant, but just to remember why exactly I did that, I've put this line back, run the `zeitwerk:check` again and all seems good. So, indeed, there is no need to put this line to the end :/ Reverted.

File: config/initializers/zeitwerk.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+Rails.autoloaders.main.ignore(
```
Comment (stejskalleos) [inline config/initializers/zeitwerk.rb:31]: @adamruzicka, @ofedoren Do you guys know what's the difference between `Rails.autoloaders.main` and ``
Rails [docs](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#rails-autoloaders) doesn't say much, after while of searching still couldn't find the difference. From first look they return same object but then why there are two methods?

```ruby
Rails.autoloaders.main
=> #<Zeitwerk::Loader:0x00000000073af3e8
 @autoloaded_dirs=[],
 @autoloads={},
 @collapse_dirs=#<Set: {}>,
 @collapse_glob_patterns=#<Set: {}>,
 @eager_load_exclusions=#<Set: {}>,
 @eager_loaded=false,
 @ignored_glob_patterns=#<Set: {}>,
 @ignored_paths=#<Set: {}>,
 @inflector=ActiveSupport::Dependencies::ZeitwerkIntegration::Inflector,
 @initialized_at=2022-07-27 15:53:58.924070552 +0200,
 @lazy_subdirs={},
 @logger=nil,
 @mutex=#<Thread::Mutex:0x00000000073aefb0>,
 @mutex2=#<Thread::Mutex:0x00000000073aef88>,
 @on_load_callbacks={},
 @on_setup_callbacks=[],
 @on_unload_callbacks={},
 @reloading_enabled=false,
 @root_dirs={},
 @setup=false,
 @tag="rails.main",
 @to_unload={}>

 Rails.autoloaders.once
=> #<Zeitwerk::Loader:0x000000000af8bd80
 @autoloaded_dirs=[],
 @autoloads={},
 @collapse_dirs=#<Set: {}>,
 @collapse_glob_patterns=#<Set: {}>,
 @eager_load_exclusions=#<Set: {}>,
 @eager_loaded=false,
 @ignored_glob_patterns=#<Set: {}>,
 @ignored_paths=#<Set: {}>,
 @inflector=ActiveSupport::Dependencies::ZeitwerkIntegration::Inflector,
 @initialized_at=2022-07-27 15:54:07.749815776 +0200,
 @lazy_subdirs={},
 @logger=nil,
 @mutex=#<Thread::Mutex:0x000000000af8b6c8>,
 @mutex2=#<Thread::Mutex:0x000000000af8b628>,
 @on_load_callbacks={},
 @on_setup_callbacks=[],
 @on_unload_callbacks={},
 @reloading_enabled=false,
 @root_dirs={},
 @setup=false,
 @tag="rails.once",
 @to_unload={}>
```

Comment (adamruzicka) [reply config/initializers/zeitwerk.rb:31]: https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#config-autoload-paths talks about which paths are managed by which autoloader

File: app/services/proxy_status/puppetca.rb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 module ProxyStatus
-  class PuppetCA < Base
+  class Puppetca < Base
```
Comment (ekohl) [inline app/services/proxy_status/puppetca.rb:2]: In https://github.com/theforeman/foreman/pull/9328 I chose to make CA an acronym and name the file `puppet_ca` instead. Didn't notice this PR until now.

File: config/initializers/zeitwerk.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  Rails.root.join('lib/core_extensions.rb'),
+  Rails.root.join('lib/generators')
+)
+Rails.autoloaders.once.ignore(
+  Rails.root.join('app/registries/foreman/access_permissions.rb'),
+  Rails.root.join('app/registries/foreman/settings.rb'),
+  Rails.root.join('app/registries/foreman/settings')
```
Comment (ekohl) [inline config/initializers/zeitwerk.rb:38]: I wondered if these shouldn't be initializers.

File: config/application.rb [hunk 8]
```diff
@@ -104,7 +104,7 @@ class Application < Rails::Application
     Dir["#{Rails.root}/config/routes/**/*.rb"].each do |route_file|
       config.paths['config/routes.rb'] << route_file
     end
-
+    config.load_defaults 6.0
```
Comment (ekohl) [inline config/application.rb:107]: I had to add
```ruby
    # Rails 5 changed this to true
    config.active_record.belongs_to_required_by_default = false
```

Comment (ekohl) [reply config/application.rb:107]: I opened https://github.com/theforeman/foreman/pull/9382 to do this as a separate preparation step.

File: app/services/setting_registry.rb (hunk 1 above)
Comment (ezr-ondrej) [reply app/services/setting_registry.rb:137]: @stejskalleos @ofedoren sorry for the delay it totally slipped my mind when I couldn't get to it that week, sorry! Here is my take on it #9524, my foreman knowledge is bit rusty so I hope I got it right.
 
If I can do some other clean-ups that would elevate some pain from this PR, I'm happy to do so, I believe it would help the project to get this done (but I know it's pain to find and fix all the blockers)
